src/linalg.jl

```diff
@@ -92,8 +81,115 @@ end
+    @safe size(xA.parent, 1) == size(C, 2) || throw(DimensionMismatch())
+    if (β != 1, ~)
+        @safe error("only β = 1 is supported, got β = $(β).")
+    end
+    @invcheckoff for col in 1:size(xA.parent, 2)
+        @inbounds for k in nzrange(xA.parent, col)
+                        @routine begin
+                            anc1 ← zero(eltype(xA.parent))
+                            anc1 += (xA.parent.nzval[k])'*α
+                        end
+            @simd for multivec_row in 1:size(X,1)
+                C[multivec_row, xA.parent.rowval[k]] += X[multivec_row, col] * anc1
+            end
+            ~@routine
+        end
+    end
+end
+
+@i function idot(r::T, A::SparseMatrixCSC{T},B::SparseMatrixCSC{T}) where {T}
+    @routine @invcheckoff begin
+        (m, n) ← size(A)
+        branch_keeper ← zeros(Bool, 2*m)
+    end
+    @safe size(B) == (m,n) || throw(DimensionMismatch("matrices must have the same dimensions"))
+    @invcheckoff @inbounds for j = 1:n
+        @routine begin
+            ia1 ← A.colptr[j]
+            ib1 ← B.colptr[j]
+            ia2 ← A.colptr[j+1]
+            ib2 ← B.colptr[j+1]
+            ia ← ia1
+            ib ← ib1
+        end
+        @inbounds for i=1:ia2-ia1+ib2-ib1-1
+            ra ← A.rowval[ia]
+            rb ← B.rowval[ib]
+            if (ra == rb, ~)
+                r += A.nzval[ia]' * B.nzval[ib]
+            end
+            ## b move -> true, a move -> false
+            branch_keeper[i] ⊻= @const ia == ia2-1 || (ib != ib2-1 && ra > rb)
+            ra → A.rowval[ia]
+            rb → B.rowval[ib]
+            if (branch_keeper[i], ~)
+                INC(ib)
+            else
+                INC(ia)
             end
         end
+        ~@inbounds for i=1:ia2-ia1+ib2-ib1-1
+            ## b move -> true, a move -> false
+            branch_keeper[i] ⊻= @const ia == ia2-1 || (ib != ib2-1 && A.rowval[ia] > B.rowval[ib])
+            if (branch_keeper[i], ~)
+                INC(ib)
+            else
+                INC(ia)
+            end
+        end
+        ~@routine
     end
+    ~@routine
 end
+
+@i function idot(r, x::AbstractVector, A::AbstractSparseMatrix{T1}, y::AbstractVector{T2}) where {T1, T2} 
+    @safe length(x) == size(A, 1) || throw(DimensionMismatch())
+    @safe length(y) == size(A, 2) || throw(DimensionMismatch())
+
+    @invcheckoff @inbounds for col in 1:size(A, 2)
+        for k in nzrange(A, col)
+            @routine begin
+                anc ← zero(promote_type(T1, T2))
+                anc += A.nzval[k] * y[col]
+            end
+            r += (x[A.rowval[k]])' * anc
+            ~@routine 
+        end
+    end
+end 
+
+@i function idot(r, x::SparseVector, A::AbstractSparseMatrix{T1}, y::SparseVector{T2}) where {T1, T2}
+    @safe length(x) == size(A, 1) || throw(DimensionMismatch())
+    @safe length(y) == size(A, 2) || throw(DimensionMismatch())
+
+    @invcheckoff for (yi, yv) in zip(y.nzind, y.nzval)
```

Comment:
Some `@inbounds` annotation maybe?

Reply:
`@inbounds` would help improve the performance.